File: src/main/java/leetcode/medium/GroupAnagrams.java
package leetcode.medium;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

//49. Group Anagrams
public class GroupAnagrams {

    public static void main(String[] args) {
        System.out.println(groupAnagrams(new String[]{"eat", "tea", "tan", "ate", "nat", "bat"}));
    }

    public static List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for (String str : strs) {
            map.computeIfAbsent(findSort(str), val -> new ArrayList<String>()).add(str);
        }
        return new ArrayList<>(map.values());
    }

    private static String findSort(String str) {
        var chars = new char[26];
        var result = new StringBuilder();
        for (char ch : str.toCharArray()) {
            chars[ch - 'a']++;
        }
        for (int i = 0; i < chars.length; i++) {
            if (chars[i] != 0) {
                result.append((char) ('a' + i)).append(chars[i]);
            }
        }
        return result.toString();
    }

}

File: src/main/java/leetcode/medium/GroupAnagrams.kt
package leetcode.medium

//49. Group Anagrams

fun main() {
    println(groupAnagrams(arrayOf("abcdefghijklmnopqrstuvwxyz", "abcdefghijklmnopqrstuvwxzy")))
    println(groupAnagrams2(arrayOf("aaaaaaaaaaaaaaaad", "daaaaaaaaaaaaaaaa")))

}

fun groupAnagrams(strs: Array<String>): List<List<String>> {
    val map = hashMapOf<String, ArrayList<String>>()
    for (s in strs) {
        val chars = IntArray(26)
        val resStr = StringBuilder()
        for (ch in s) {
            chars[ch - 'a']++
        }
        for (i in 0 until 26) {
            if (chars[i] > 0)
                resStr.append('a' + i.toString().repeat(chars[i]))
        }
        if (map.containsKey(resStr.toString())) {
            map[resStr.toString()]!!.add(s)
        } else {
            map[resStr.toString()] = arrayListOf(s)
        }
    }
    return ArrayList(map.values)
}

fun groupAnagrams2(strs: Array<String>): List<List<String>> {
    val map = HashMap<String, ArrayList<String>>()
    for (s in strs) {
        val chars = s.toCharArray()
        chars.sort()
        val sortedString = String(chars)
        if (map.containsKey(sortedString)) {
            map[sortedString]!!.add(s)
        } else {
            map[sortedString] = arrayListOf(s)
        }
    }
    return ArrayList(map.values)
}